docs/core/v3.0.rst

```diff
@@ -0,0 +1,1763 @@
+possibilities that remain to be explored. A goal of this specification
+is to define a format with a number of clear extension points and
+mechanisms, in order to provide a framework for freely building on and
+exploring these possibilities. We aim to make this possible, whilst
+also providing pathways for a graceful degradation of functionality
+where possible, in order to retain interoperability. We also aim to
+provide a framework for community-defined extensions, which can be
+developed and published independently without requiring centralised
+coordination of all specifications.
+
+See `extensions`_ below.
+
+Stability Policy
+----------------
+
+This core specification adheres to a ``MAJOR.MINOR`` version
+number format. A zarr implementation provides the read and write API by
+implementing this specification can be considered compatible with all
+datasets following the specification with the same major version number.
+
+Notably, this excludes extensions, codecs and stores from the compatibility
+of the core specification. However, versioned extensions and stores are also
+expected to follow this stability policy.
+
+This means that implementations based on a ``3.X`` specification will be
+able to read and write to datasets that follow any ``3.Y`` specification,
+as long as it only uses
+
+- optional extensions or those supported by the implementation,
+- no codec or one supported by the implementation, and
+- the used store must be supported by the implementation.
+
+For details, please see the `zarr_format`_ metadata entry.
+
+
+Questions that still need to be resolved
+----------------------------------------
+
+We solicit feedback on the following area during the RFC period of this first
+draft.
+
+ - Should core metadata and user attributes be stored together or separate documents?
+   (See https://github.com/zarr-developers/zarr-specs/issues/72)
+ - extensions and ``must_understand = True`` might be too restrictive. Work a
+   We propose to develop a draft implementation with extensions and
+   see how far we can go. A possible list of extensions to include:
+
+    - Boolean
+    - Complex
+    - Datetime
+    - Named dimensions
+    - Awkward arrays
+
+   See https://github.com/zarr-developers/zarr-specs/issues/89 for discussion on
+   the topic.
+
+  - Node name case sensitivity: The node name is now case sensitive. This may
+    make store implementation more complicated as some backends might not be
+    (like some specific filesystem / object store), and we may want to
+    recommend a standard escaping mechanism in those cases.
+    https://github.com/zarr-developers/zarr-specs/issues/57
+
+  - Node name character set: We
+    solicit feedback on whether store implementation should support full unicode.
+    https://github.com/zarr-developers/zarr-specs/issues/56
+
+  - Should named dimensions be part of the core metadata spec?
+    https://github.com/zarr-developers/zarr-specs/issues/73
```

Comment:
I think basic support for named dimensions would be a win, because it would support consistent use of this metadata across domains. I would not go any more complex than an optional JSON array of strings, though, whose length is required to match the array dimensionality.

Reply:
No objection from me. Can I suggest that if anyone has an objection to this, please add a thumbs down or add a comment in this conversation.

Reply:
I'd object, but I'm not strongly against it:

In `anndata` we store things that are like arrays, but are not dense ndarrays. Currently this is tables and sparse arrays, but will soon include ragged arrays. These arrays share logical dimensions with dense arrays, but are stored as a collection of physical arrays in a group. If named dimensions are metadata on zarr.Arrays, then we wouldn't be able to use this system on our other array types. We end up needing our own system for this, so would probably just opt out of the whole native dimension system so we can handle dimensions consistently.

One other example would be the coordinates and dimensions specifications for OME-NGFF. We need much more info about our dimensions that just a name, so we will encode it separately. If dimension info is already stored in the metadata, I'd argue we shouldn't also store it in `.zarray` since we'd be responsible for keeping it in sync. 

I suspect this will be the case for many fields, so different domains are likely to use other solutions. There is probably a common concept of "array dimensions" across fields, but I think it's going to be more complicated. More on the netcdf, cf-conventions level than the array storage library. Having a solution simple enough for some domains, but not powerful enough for others, may even harm adoption of a common standard that solves both cases.

All that said: this adds little complexity and is easy to opt out of. So the main argument here is: I think the value is limited, and it'd be nice to keep the spec as lean as possible.

Reply:
@ivirshup 

My understanding is that the proposal here is for there to be a mechanism to add optional dimension names to zarr, which would presumably be supported by zarr-python and other zarr implementations, e.g. via `array.dimension_names` or similar syntax.   If it is a core feature, then we would have something like:

(a) array metadata document:

```json
{
  "data_type": ...,
  "shape": [100, 200, 300],
  "dimension_labels": ["x", "y", "z"],
  "attributes": { ... },
  ...
}
```

If it is an extension, then we would have something like:


(b) array metadata document:

```json
{
  "data_type": ...,
  "shape": [100, 200, 300],
  "attributes": { ... },
  "extensions": [
    {"extension": "https://purl.org/zarr/spec/dimension_labels",
     "must_understand": false,
     "configuration": {
        "dimension_labels": ["x", "y", "z"]
     }
    }
  ]
  ...
}
```

But the zarr-python API would be exactly the same either way and there would be little practical difference; the metadata representation would simply be more verbose in the extension case.

To be clear, are you saying that you would prefer that zarr-python not support named dimensions at all, or that you prefer option (b)?

You raise two important cases:

1. There is a real zarr array, with meaningful dimension names, but we may want to store additional per-dimension information like units, coordinate array, etc. in addition to the dimension name.  This is the case for an OME-Zarr array where we have unit information, or xarray and netcdf where we have coordinate arrays.
2. Some software or specification cares about dimension names, and deals with both zarr arrays, and other things that aren't zarr arrays but are encoded as zarr group attributes and/or a collection of zarr arrays, and potentially have meaningful dimension names.  This is the case for OME multiscale datasets, and for anndata and ragged arrays.

For case 1, I would say it would still be very useful for zarr-python or other zarr implementations to be able to interpret the dimension names, even without being able to interpret the additional per-dimension information like units or coordinate arrays.  I don't think there is any particular difficulty in storing the units or other information as an attribute separate from the dimension labels.  I agree that duplicating the dimension labels would be a bit unfortunate since it may be confusing if they get out of sync.

For case 2, it is true that if zarr has its own dimension label mechanism, then a completely separate mechanism would be needed to add dimension labels for things that aren't zarr arrays.  But these other things could still specify their dimension labels in a similar way, e.g. with a `dimension_labels` attribute.

Reply:
It would be great if you could move this discussion into the more concrete issues and PRs, namely #73 and #162 (#144 is also relevant, however this is mixed with other issues). The plan is to resolve or migrate all ongoing threads here since it became quite full. Sorry for the inconvenience 🫤